tiledb/sm/subarray/cell_slab_iter.cc

```diff
@@ -0,0 +1,324 @@
+  T tile_end = (range[1] - dim_domain_start) / tile_extent;
+
+  // The range falls int he same tile
+  if (tile_start == tile_end) {
+    ranges->emplace_back(range[0], range[1], tile_start);
+  } else {  // We need to split the range
+    T start = range[0];
+    T end;
+    for (T i = tile_start; i < tile_end; ++i) {
+      end = (i + 1) * tile_extent + dim_domain_start - 1;
+      ranges->emplace_back(start, end, i);
+      start = end + 1;
+    }
+    ranges->emplace_back(start, range[1], tile_end);
+  }
+}
+
+template <class T>
+void CellSlabIter<T>::init_cell_slab_lengths() {
+  auto layout = subarray_->layout();
+  auto dim_num = subarray_->dim_num();
+
+  if (layout == Layout::ROW_MAJOR) {
+    auto range_num = ranges_[dim_num - 1].size();
+    cell_slab_lengths_.resize(range_num);
+    for (size_t i = 0; i < range_num; ++i)
+      cell_slab_lengths_[i] =
+          ranges_[dim_num - 1][i].end_ - ranges_[dim_num - 1][i].start_ + 1;
+  } else {
+    assert(layout == Layout::COL_MAJOR);
+    auto range_num = ranges_[0].size();
+    cell_slab_lengths_.resize(range_num);
+    for (size_t i = 0; i < range_num; ++i)
+      cell_slab_lengths_[i] = ranges_[0][i].end_ - ranges_[0][i].start_ + 1;
+  }
+}
+
+template <class T>
+void CellSlabIter<T>::init_coords() {
+  auto dim_num = subarray_->dim_num();
+
+  range_coords_.resize(dim_num);
+  cell_slab_coords_.resize(dim_num);
+  for (unsigned i = 0; i < dim_num; ++i) {
+    range_coords_[i] = 0;
+    cell_slab_coords_[i] = ranges_[i][0].start_;
+  }
+}
+
+template <class T>
+Status CellSlabIter<T>::init_ranges() {
+  // For easy reference
+  auto dim_num = subarray_->dim_num();
+  auto array_schema = subarray_->array()->array_schema();
+  auto array_domain = (T*)array_schema->domain()->domain();
+  auto tile_extents = (T*)array_schema->domain()->tile_extents();
+  uint64_t range_num;
+  const T* range;
+  T tile_extent, dim_domain_start;
+
+  ranges_.resize(dim_num);
+  for (unsigned i = 0; i < dim_num; ++i) {
+    RETURN_NOT_OK(subarray_->get_range_num(i, &range_num));
+    ranges_[i].reserve(range_num);
+    tile_extent = tile_extents[i];
+    dim_domain_start = array_domain[2 * i];
+    for (uint64_t j = 0; j < range_num; ++j) {
+      RETURN_NOT_OK(subarray_->get_range(i, j, (const void**)&range));
+      create_ranges(range, tile_extent, dim_domain_start, &ranges_[i]);
+    }
+  }
+
+  return Status::Ok();
+}
+
+template <class T>
+Status CellSlabIter<T>::sanity_check() const {
+  assert(subarray_ != nullptr);
+
+  // Check layout
+  auto layout = subarray_->layout();
+  if (layout != Layout::ROW_MAJOR && layout != Layout::COL_MAJOR)
+    return LOG_STATUS(Status::CellSlabIterError(
+        "Unsupported subarray layout; the iterator supports only row-major and "
+        "column-major layouts"));
+
+  // Check type
+  bool error;
+  switch (subarray_->type()) {
+    case Datatype::INT8:
+      error = &typeid(T) != &typeid(int8_t);
```

Comment:
If you want you can use `std::is_same<T, int8_t>()::value` here for comparing types, might be a little cleaner.

Reply:
Agreed, I will modify as suggested.